hadoop-common-project/hadoop-common/src/main/java/io/hops/security/UsersGroupsCache.java

```diff
@@ -15,289 +15,786 @@
+        return 0;
+      }
+      throw new IOException(e);
     }
-    cleanCacheOnGroupRemoval(groupName, removeUsers);
   }
-
-  private void removeGroup(Integer groupId, boolean removeUsers){
-    if(groupId == null)
+  
+  Integer getGroupIdFromCache(String groupName){
+    if(groupName == null)
+      return 0;
+    return groupsToIds.getIfPresent(groupName);
+  }
+  
+  String getGroupName(Integer groupId) throws IOException {
+    if(!isConfigured)
+      return null;
+    
+    if(groupId == null || groupId <= 0)
+      return null;
+    try {
+      return idsToGroups.get(groupId);
+    } catch (ExecutionException e) {
+      if(e.getCause() instanceof GroupNotFoundException){
+        return null;
+      }
+      throw new IOException(e);
+    }
+  }
+  
+  
+  void removeUserFromGroup(String userName, String groupName) throws IOException {
+    if(!isConfigured)
       return;
-
-    String groupName = idsToGroups.remove(groupId);
-    if(groupName != null){
-      groupsToIds.remove(groupName);
-      cleanCacheOnGroupRemoval(groupName, removeUsers);
+    
+    if(userName == null || groupName == null)
+      return;
+    
+    LOG.debug("Remove user-group from DB user=" + userName + ", group=" + groupName);
+    Integer userId = getUserId(userName);
+    Integer groupId = getGroupId(groupName);
+  
+    removeUserFromGroupFromDB(userId, groupId);
+    removeUserFromGroupInCache(userName, groupName);
+  }
+  
+  List<String> getGroups(String user) throws IOException {
+    if(!isConfigured)
+      return null;
+    
+    if(user == null)
+      return null;
+    try {
+      return usersToGroups.get(user);
+    } catch (ExecutionException e) {
+      if(e.getCause() instanceof GroupsNotFoundForUserException){
+        return null;
+      }
+      throw new IOException(e);
     }
   }
-
-  private void cleanCacheOnUserRemoval(String userName){
-    cleanCacheOnUserRemoval(userName, usersToGroups.remove(userName));
+  
+  List<String> getGroupsFromCache(String user) throws IOException {
+    if (user == null)
+      return null;
+    return usersToGroups.getIfPresent(user);
   }
-
-  private void cleanCacheOnUserRemoval(String userName, List<String> groups){
-    if(groups == null)
+  
+  
+  void removeUserFromGroupTx(String user, String group, boolean cacheOnly) throws IOException{
+    if(cacheOnly){
+      if(user != null && group == null){
+        removeUserFromCache(user);
+      }else if(user == null && group != null){
+        removeGroupFromCache(group);
+      }else if(user != null && group != null){
+        removeUserFromGroupInCache(user, group);
+      }
+    }else{
+      if(user != null && group == null){
+        removeUser(user);
+      }else if(user == null && group != null){
+        removeGroup(group);
+      }else if(user != null && group != null){
+        removeUserFromGroup(user, group);
+      }
+    }
+  }
+  
+  void addUserToGroupTx(String user, String group, boolean cacheOnly) throws IOException{
```

Comment:
I find cacheOnly name kind of confusing. If I set it to false, I would expect it to add both in cache and DB which is not the case.

Reply:
If cacheOnly is set to false, it adds to the DB as well as the cache. Adding to the cache is done implicitly, you can check the methods addUser and addGroup and addUserToGroup